bitbybit/src/bitenum.rs

```diff
@@ -260,16 +433,54 @@ pub fn bitenum(args: TokenStream, input: TokenStream) -> TokenStream {
         )
     };
 
+    let wrapped_variants = variants
+        .iter()
+        .map(|variant| {
+            BitenumVariant {
+                variant,
+                // If enum has both variants with fields and ones with discriminants, `#[repr]`
+                // must be specified. We shouldn't emit that because user code may already specify
+                // it. Work around this for bitenums with fields by not emitting the discrimants.
+                // This shouldn't matter because "as" already won't work for these non-primitive types.
+                emit_discriminant: !has_ranges && !has_catchall
+            }
+        })
+        .collect::<Punctuated<BitenumVariant, Comma>>();
+
+    let raw_value_function = {
+        let raw_value_cases: Punctuated<TokenStream2, Comma> = emitted_variants.iter().map(|(value, name, cfg_attributes)| {
+            match value {
+                VariantValue::SingleValue(expression, _) => quote! {
+                    #( #cfg_attributes )*
+                    Self::#name => #result_constructor(#expression)
+                },
+                VariantValue::Ranges(_) | VariantValue::Catchall => quote! {
+                    #( #cfg_attributes )*
+                    Self::#name(v) => v
+                },
+            }
+        })
+        .collect();
+
+        quote! {
+            /// Returns the underlying raw value of this bitfield
+            pub const fn raw_value(self) -> #bounded_data_type {
+                match self {
+                    #raw_value_cases
+                }
+            }
+        }
+    };
+
     let expanded = quote! {
         #[derive(Copy, Clone)]
         #( #enum_attrs )*
         #enum_vis enum #enum_name {
-            #variants
+            #wrapped_variants
         }
 
         impl #enum_name {
-            /// Returns the underlying raw value of this bitfield
-            pub const fn raw_value(self) -> #bounded_data_type { #result_constructor(self as #base_data_type) }
+            #raw_value_function
```

Comment:
did you verify this is all removed by the compiler for simple enums? I wonder if maybe we should use the old-codepath if the new features aren't used? let me know if it's unclear how to test - I can help with that

Reply:
It looks like the compiler (in release mode) emits roughly the same assembly in both cases. I'm not 100% sure I'm looking correctly though. We can discuss next week.

Reply:
Then again, it's easy enough to just keep the old code path, so I'll just do that.